src/nbl/asset/utils/CIESProfile.cpp

```diff
@@ -0,0 +1,300 @@
+    T expected = obj->load();
+    while (!std::atomic_compare_exchange_weak(obj, &expected, expected + arg));
+
+    return expected;
+}
+
+template<class ExecutionPolicy>
+core::smart_refctd_ptr<asset::ICPUImageView> CIESProfile::createIESTexture(ExecutionPolicy&& policy, const size_t& width, const size_t& height) const
+{
+    if constexpr (policy._Parallelize)
+        static_assert(false, "meh it appears that filters with par policy are broken on this branch, use seq policy till fixed");
+
+    asset::ICPUImage::SCreationParams imgInfo;
+    imgInfo.type = asset::ICPUImage::ET_2D;
+    imgInfo.extent.width = width;
+    imgInfo.extent.height = height;
+    imgInfo.extent.depth = 1u;
+    imgInfo.mipLevels = 1u;
+    imgInfo.arrayLayers = 1u;
+    imgInfo.samples = asset::ICPUImage::ESCF_1_BIT;
+    imgInfo.flags = static_cast<asset::IImage::E_CREATE_FLAGS>(0u);
+    imgInfo.format = IES_TEXTURE_STORAGE_FORMAT;
+    auto outImg = asset::ICPUImage::create(std::move(imgInfo));
+
+    asset::ICPUImage::SBufferCopy region;
+    constexpr auto texelBytesz = asset::getTexelOrBlockBytesize<IES_TEXTURE_STORAGE_FORMAT>();
+    const size_t bufferRowLength = asset::IImageAssetHandlerBase::calcPitchInBlocks(width, texelBytesz);
+    region.bufferRowLength = bufferRowLength;
+    region.imageExtent = imgInfo.extent;
+    region.imageSubresource.baseArrayLayer = 0u;
+    region.imageSubresource.layerCount = 1u;
+    region.imageSubresource.mipLevel = 0u;
+    region.bufferImageHeight = 0u;
+    region.bufferOffset = 0u;
+
+    auto buffer = core::make_smart_refctd_ptr<asset::ICPUBuffer>(texelBytesz * bufferRowLength * height);
+
+    if (!outImg->setBufferAndRegions(std::move(buffer), core::make_refctd_dynamic_array<core::smart_refctd_dynamic_array<asset::IImage::SBufferCopy>>(1ull, region)))
+        return {};
+
+    //! Generate 2D IES grid data CandelaPower Distribution Curve texture can be created from
+    {
+        const auto& creationParams = outImg->getCreationParameters();
+
+        CFillImageFilter::state_type state;
+        state.outImage = outImg.get();
+        state.subresource.aspectMask = static_cast<IImage::E_ASPECT_FLAGS>(0);
+        state.subresource.baseArrayLayer = 0u;
+        state.subresource.layerCount = 1u;
+        state.outRange.extent = creationParams.extent;
+
+        const IImageFilter::IState::ColorValue::WriteMemoryInfo wInfo(creationParams.format, outImg->getBuffer()->getPointer());
+
+        const double maxValue = getMaxValue();
+        const double maxValueRecip = 1.0 / maxValue;
+
+        const double vertInv = 1.0 / height;
+        const double horiInv = 1.0 / width;
+
+        const double dTheta = core::PI<double>() / height;
+        const double dPhi = 2 * core::PI<double>() / width;
+
+        integral = 0.0;
+        size_t nonZeroEmissionDomainSize = 0u;
+
+        std::atomic<IES_STORAGE_FORMAT> aIntegral = 0.0;
+        std::atomic<size_t> aNonZeroEmissionDomainSize = 0;
```

Comment:
if you use atomics, you'll get negligible benefit from parallelization because of atomic contention.